consensus/src/lib.rs

```diff
@@ -1,27 +1,1112 @@
+        };
+        let tx_index = tx_db.get_mainchain_tx_index(&tx_id)?.ok_or(BlockError::Unknown)?;
+        match tx_index.get_tx_position() {
+            SpendablePosition::Transaction(position) => {
+                tx_db.get_mainchain_tx_by_position(position)?.ok_or(BlockError::Unknown)
+            }
+            SpendablePosition::BlockReward(_) => unimplemented!(),
+        }
+    }
+
+    fn get_output_by_outpoint(
+        tx_db: &<blockchain_storage::Store as Transactional<'a>>::TransactionRw,
+        outpoint: &OutPoint,
+    ) -> Result<common::chain::TxOutput, BlockError> {
+        let tx = Self::get_tx_by_outpoint(tx_db, outpoint)?;
+        let output_index = outpoint.get_output_index() as usize;
+        assert!(output_index <= tx.get_outputs().len());
+        Ok(tx.get_outputs()[output_index].clone())
+    }
+
+    fn get_input_value(
+        tx_db: &<blockchain_storage::Store as Transactional<'a>>::TransactionRw,
+        input: &common::chain::TxInput,
+    ) -> Result<Amount, BlockError> {
+        let tx = Self::get_tx_by_outpoint(tx_db, input.get_outpoint())?;
+        let output_index = input.get_outpoint().get_output_index() as usize;
+        assert!(output_index <= tx.get_outputs().len());
+        Ok(tx.get_outputs()[output_index].get_value())
+    }
+
+    fn check_block_fee(&self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        let input_mlt = transactions
+            .iter()
+            .map(|x| {
+                x.get_inputs()
+                    .iter()
+                    .map(|input| {
+                        Self::get_input_value(&self.db_tx, input).expect("Couldn't get input")
+                    })
+                    .sum::<Amount>()
+            })
+            .sum();
+        let output_mlt: Amount = transactions
+            .iter()
+            .map(|x| x.get_outputs().iter().map(|output| output.get_value()).sum::<Amount>())
+            .sum();
+
+        // Check that fee is not negative
+        if output_mlt >= input_mlt {
+            return Err(BlockError::Unknown);
+        }
+        Ok(())
+    }
+
+    fn check_tx_inputs(&self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        let mut total_value = Amount::new(0);
+        for tx in transactions {
+            for input in tx.get_inputs() {
+                let tx_index =
+                    Self::get_mainchain_index_by_outpoint(&self.db_tx, input.get_outpoint())?;
+                // If there is a wrong input index then it's cause a BlockError
+                let output = Self::get_output_by_outpoint(&self.db_tx, input.get_outpoint())?;
+
+                // Check is input has already spent
+                if tx_index
+                    .get_spent_state(input.get_outpoint().get_output_index())
+                    .map_err(BlockError::from)?
+                    == OutputSpentState::Unspent
+                {
+                    return Err(BlockError::Unknown);
+                }
```

Comment:
Unspent input is an error? Am I reading this incorrectly?

Reply:
I'm concerned... if this is a bug, how was it not caught in unit tests.

Reply:
If we connect a block, transaction inputs should be already spent, isn't that?

We are far away to full cover the code with unit tests here, I'm still working on it.

Reply:
When you want to connect a block:

- Inputs are not spent
- Then you connect the block, which spends these inputs
- After the success of the connection, he inputs are spent.

Another thing:
And please refrain from using vague function names. "check_tx_inputs" doesn't really mean anything. I'd rather read a function like "check_tx_inputs_unspent" or something like that. I understand the language barrier, but please push with this a little bit. We're supposed to mostly understand what the function does from its name.

Reply:
`check_tx_inputs`— unfortunately, at this time it is not a language barrier. This naming appear due to I know that there should check inputs, but not certain which ones, I imagine that there should be a bunch of checks but don't really know what it should be. And this is a reason why I choose an abstract name. But anyway, this is a good point, I will try to avoid abstract naming in the future. Thank you!

Reply:
```
When you want to connect a block:

Inputs are not spent
Then you connect the block, which spends these inputs
After the success of the connection, the inputs are spent.
```
Added the test - `test_spend_inputs_simple` and it's work exactly as you said. After the success of the connection, the inputs are spent. Probably, I misunderstood something, but seems its work fine. Continue investigate it.